### PR TITLE
3547 Interceptor hook for client-assigned IDs

### LINKED ISSUE #3547: Interceptor hook for client-assigned IDs
**Describe the issue**
We have client-assigned IDs disabled on our server because we want to prevent customers from creating resources that conflict with each other. However, there have been a variety of occasions where it would be desirable to set a client-assigned ID.
Right now it is possible to allow client-assigned IDs on some resource and not others by using multiple persistence modules. However, we would like more flexibility so that we can decide whether not a certain user is authorized to create client-assigned IDs.

**Solution design**
In this case we think the most flexible option would be to have an interceptor that decides whether creation can proceed when the client provides resource ID. We want to be able to decide this based on the user sending the request.
- Add new void Pointcut just before client-assigned id is created so can allow/deny based on `ResourceDetails` and `RequestDetails`

### PR DESCRIPTION
**Pull request by jdar8**
closes #3547 

- Added new pointcut that takes in `ResourceDetails` and `RequestDetails`
- Added test
- Added changelog

### CODE REVIEW

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/interceptor/api/Pointcut.java [hunk 1]
```diff
@@ -1345,6 +1345,34 @@ public enum Pointcut implements IPointcut {
...
+	 * pulled out of the servlet request. Note that the bean
+	 * properties are not all guaranteed to be populated, depending on how early during processing the
+	 * exception occurred.
+	 * </li>
+	 * </ul>
+	 * <p>
+	 * Hooks should return <code>void</code>.
+	 * </p>
+	 */
+	STORAGE_PRESTORAGE_ACCEPT_CLIENT_ASSIGNED_ID(void.class,
```
Comment (tadgh) [inline hapi-fhir-base/src/main/java/ca/uhn/fhir/interceptor/api/Pointcut.java:1371]: This name should probably be `STORAGE_PRESTORAGE_CLIENT_ASSIGNED_ID`. The implementers may or may not actually perform any acceptance checking.

File: hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/6_0_0/3547-interceptor-hook-for-client-assigned-ids.yaml [hunk 2]
```diff
@@ -0,0 +1,5 @@
+type: add
+issue: 3547
+jira: SMILE-4141
+title: "Added a pointcut just before the client assigned id is created. This allows interceptors to allow or deny the
```
Comment (tadgh) [inline hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/6_0_0/3547-interceptor-hook-for-client-assigned-ids.yaml:4]:
```suggestion
title: "Added a new pointcut: `STORAGE_PRESTORAGE_CLIENT_ASSIGNED_ID`  which is invoked when a user attempts to create a resource with a client-assigned ID. 
```

File: hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4InterceptorTest.java [hunk 3]
```diff
@@ -511,13 +514,33 @@ public void resourcePreCreate(IBaseResource theResource) {
...
+		mySrdInterceptorService.registerInterceptor(interceptor);
+
+		Patient p = new Patient();
+		try {
+			myPatientDao.create(p, mySrd).getId().toUnqualifiedVersionless();
+			fail();
+		} catch (ForbiddenOperationException e) {
+			assertEquals("Client assigned id rejected.", e.getMessage());
+		}
+	}
```
Comment (tadgh) [inline hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4InterceptorTest.java:542]: Issue: This seems wrong. Calling `.create()` on the dao will _always_ use a server assigned ID. I would expect this call to _not_ invoke your interceptor. I would expect your interceptor to be invoked on `.update()` calls to the dao _only_.

Comment (tadgh) [reply hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4InterceptorTest.java:542]: When you call .create(), does it actually _use_ your client assigned ID?

**Review by tadgh — CHANGES_REQUESTED** [review]

File: hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4InterceptorTest.java (hunk 3 above)
Comment (jdar8) [reply hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4InterceptorTest.java:553]: yea it is wrong. create sets it to an empty string

fixed

**Review by tadgh — APPROVED** [review]

**Pull request merged**